osgtest/tests/test_155_stashcache.py

```diff
@@ -0,0 +1,157 @@
+import os
+import pwd
+
+from osgtest.library import core
+from osgtest.library import files
+from osgtest.library.osgunittest import OSGTestCase
+from osgtest.library import service
+
+
+CACHE_DIR = "/tmp/sccache"
+CACHE_XROOT_PORT = 1094  # can't change this - stashcp doesn't allow you to specify port
+CACHE_HTTP_PORT = 8001
+ORIGIN_XROOT_PORT = 1095
+ORIGIN_DIR = "/tmp/scorigin"
+CACHE_AUTHFILE_PATH = "/etc/xrootd/Authfile-cache"
+CACHE_CONFIG_PATH = "/etc/xrootd/xrootd-stashcache-cache-server.cfg"
+ORIGIN_CONFIG_PATH = "/etc/xrootd/xrootd-stashcache-origin-server.cfg"
+CACHES_JSON_PATH = "/etc/stashcache/caches.json"
+
+
+# TODO Set up authenticated stashcache as well
+CACHE_CONFIG_TEXT = """\
+all.export  /
+set cachedir = {CACHE_DIR}
+xrd.allow host *
+sec.protocol  host
+all.adminpath /var/spool/xrootd
+
+xrootd.trace emsg login stall redirect
+ofs.trace all
+xrd.trace all
+cms.trace all
+
+ofs.osslib  libXrdPss.so
+# normally this is the redirector but we don't have one in this environment
+pss.origin localhost:{ORIGIN_XROOT_PORT}
+pss.cachelib libXrdFileCache.so
+pss.setopt DebugLevel 1
+
+oss.localroot $(cachedir)
+
+pfc.blocksize 512k
+pfc.ram       1024m
+# ^ xrootd won't start without a gig
+pfc.prefetch  10
+pfc.diskusage 0.90 0.95
+
+ofs.authorize 1
+acc.audit deny grant
+
+acc.authdb {CACHE_AUTHFILE_PATH}
+sec.protbind * none
+xrd.protocol http:{CACHE_HTTP_PORT} libXrdHttp.so
+
+xrd.port {CACHE_XROOT_PORT}
+
+http.listingdeny yes
+http.staticpreload http://static/robots.txt /etc/xrootd/stashcache-robots.txt
+
+# Tune the client timeouts to more aggressively timeout.
+pss.setopt ParallelEvtLoop 10
+pss.setopt RequestTimeout 25
+#pss.setopt TimeoutResolution 1
+pss.setopt ConnectTimeout 25
+pss.setopt ConnectionRetry 2
+#pss.setopt StreamTimeout 35
+
+all.sitename osgtest
+
+xrootd.diglib * /etc/xrootd/digauth.cf
+""".format(**globals())
+
+
+CACHE_AUTHFILE_TEXT = """\
+u * / rl
+"""
+
+
+ORIGIN_CONFIG_TEXT = """\
+xrd.allow host *
+sec.protocol  host
+sec.protbind  * none
+all.adminpath /var/spool/xrootd
+all.pidpath /var/run/xrootd
+
+# The directory on local disk containing the files to share, e.g. "/stash".
+oss.localroot {ORIGIN_DIR}
+all.export /
+
+xrd.port {ORIGIN_XROOT_PORT}
+all.role server
+
+xrootd.trace emsg login stall redirect
+ofs.trace all
+xrd.trace all
+cms.trace all
+""".format(**globals())
+
+
+CACHES_JSON_TEXT = """\
+[
+{"name":"root://localhost", "status":1, "longitude":-89.4012, "latitude":43.0731}
+]
+"""
```

Comment:
Wow this is a lot of config. Can we simplify this after https://github.com/opensciencegrid/StashCache-Daemon/pull/21 gets released?

Reply:
Slightly. If we start with the cache server config from that PR (and it doesn't look like it has any origin server config changes), I'd need to make the following config changes as part of setup:

- set `cachedir` and `sitename` in `10-osg-site-local.cfg`
- change the HTTPS port in `40-osg-http.cfg`
- change the HTTP port in `40-osg-http.cfg`
- delete `40-osg-monitoring.cfg`
- change both `pss.origin` settings in `40-osg-xcache.cfg`

So basically we'd be looking at a bunch of `sed`-equivalents instead of one config file write. I suppose that's probably the right thing to do in the long run anyway...

Reply:
Would we be able to instead write a `99-local.cfg` that overrides the values in the files you mentioned or does XRootD config not work that way?

Reply:
That's a good question. @bbockelm, if there are duplicate settings, do the later lines override the earlier ones?

Reply:
Depends on the configuration variable.

Xrootd configs are not a hash lookup (as with an INI file or HTCondor config) but rather linear parsing.  So, some attributes are additive (IIRC, the `all.export` lines are an example of this) where, with others, the last-specified wins.

I think everything you mention above is "last-specified wins".

Not clear - why does `40-osg-monitoring.cfg` need to be deleted though?  For the unit tests, I'd be really happy if we can stick to changing the configs in ways that we recommend to admins...

Reply:
In that case, can you add a `-itb` suffix to the monitoring hostnames?  I don't think there's something listening on UDP there right now, but I'd like to have these show up eventually (as opposed to dropping them...)

Reply:
Be careful what you wish for :P

So the new tasks will be:

- set `cachedir` and `sitename` in `10-osg-site-local.cfg`
- change the HTTPS port in `40-osg-http.cfg`
- change the HTTP port in `40-osg-http.cfg`
- replace xrd-*.osgstorage.org with xrd-*-itb.osgstorage.org in `40-osg-monitoring.cfg`
- change both `pss.origin` settings in `40-osg-xcache.cfg`

Reply:
Only the first bullet is correct - for the others, we should be able to do this solely via a 99-overrides.cfg

Reply:
Only the first bullet is correct - for the others, we should be able to do this solely via a 99-overrides.cfg

Reply:
OK. Well, we'll do that for the next version of the tests.